rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -172,18 +187,258 @@ impl From<i32> for CheckedWeight {
     }
 }
 
+impl From<i8> for CheckedWeight {
+    fn from(item: i8) -> Self {
+        Self { value: item as i32 }
+    }
+}
+
 impl From<CheckedWeight> for i64 {
     fn from(item: CheckedWeight) -> Self {
         item.value as i64
     }
 }
 
+impl From<&CheckedWeight> for i64 {
+    fn from(item: &CheckedWeight) -> Self {
+        item.value as i64
+    }
+}
+
+// Unbounded weights store essentially an unbounded integer, but are
+// optimized for small weights.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Clone)]
+pub struct UnboundedWeight {
+    pub value: Either<i32, BigInt>,
+}
```

Comment:
It's not about just size, it's about speed too. If adding 128 bits is slower than adding 32 or 64, then I would rather use the smaller size. In C++ this would be a union with a pointer and an int discriminated by a bool, so it should take much less space; (some implementations may even use the top bit of the pointer for the bool). 25 bytes is a lot for a variant which should almost never be used.

Reply:
This is what Either is supposed to do. I will use my own enum, but I don't really see what the problem of Either is.

Reply:
It's an unnecessary dependency